[EC-377] Transition Policy service into providing observables

r-tome opened the pull request:
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [X] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

<!--Describe what the purpose of this PR is. For example: what bug you're fixing or what new feature you're adding-->

## Code changes

<!--Explain the changes you've made to each file or major component. This should help the reviewer understand your changes-->
<!--Also refer to any related changes or PRs in other repositories-->

- **file.ext:** Description of what was changed and why

## Before you submit

<!-- (mark with an `X`) -->

```
- [X] I have checked for **linting** errors (`npm run lint`) (required)
- [X] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)
```

MGibson1 commented on libs/common/src/services/policy/policy.service.ts:
```diff
@@ -13,32 +17,37 @@ import { ListResponse } from "../../models/response/listResponse";
+
+  async getAll(type?: PolicyType): Promise<Policy[]> {
+    const policies = this._policies.getValue();
+
     if (type != null) {
-      return response.filter((policy) => policy.type === type);
+      return policies.filter((policy) => policy.type === type);
     } else {
-      return response;
+      return policies;
     }
   }
```
This method is not in line with the reactive way we want to be accessing data. The right way to do this is to either do this filtering on the consumer-side, or, if you want a helper in the Service, to use [`map`](https://rxjs.dev/api/index/function/map).

```typescript
  getAll(type?: PolicyType): Observable<Policy[]> {
    if (type == null) {
      return this.policies$;
    }
    return this.policies$.pipe(map((policies) => policies.filter((policy) => policy.type === type)));
  }
```

Then the consumers of this function will be able to subscribe to only the policies they're interested in

r-tome replied:
I've pushed changes where I've deleted the `getAll` method and updated the components to subscribe to the policies observable collection. In some places I had to use `firstValueFrom`.

MGibson1 reviewed (changes requested):
In general, each method in the PolicyService class needs to be rethought like I outlined for `getAll`. Think through how to make them a stream of values rather than ones grabbed ad-hoc as you need them and update usages to expect the same.

MGibson1 commented on libs/angular/src/components/settings/vault-timeout-input.component.ts:
```diff
@@ -50,21 +54,32 @@ export class VaultTimeoutInputComponent implements ControlValueAccessor, Validat
+          if (
+            await this.policyService.policyAppliesToUser(policies, PolicyType.MaximumVaultTimeout)
+          ) {
+            const vaultTimeoutPolicy = policies.filter(
+              (policy) => policy.enabled && policy.type === PolicyType.MaximumVaultTimeout
+            );
+
+            this.vaultTimeoutPolicy = vaultTimeoutPolicy[0];
+            this.vaultTimeoutPolicyHours = Math.floor(this.vaultTimeoutPolicy.data.minutes / 60);
+            this.vaultTimeoutPolicyMinutes = this.vaultTimeoutPolicy.data.minutes % 60;
+
+            this.vaultTimeouts = this.vaultTimeouts.filter(
+              (t) =>
+                t.value <= this.vaultTimeoutPolicy.data.minutes &&
+                (t.value > 0 || t.value === VaultTimeoutInputComponent.CUSTOM_VALUE) &&
+                t.value != null
+            );
+            this.validatorChange();
```
Please extract this to a private `applyVaultTimeoutPolicy` method

MGibson1 commented on libs/common/src/services/passwordGeneration.service.ts:
```diff
@@ -258,7 +259,9 @@ export class PasswordGenerationService implements PasswordGenerationServiceAbstr
       this.policyService == null
         ? null
-        : await this.policyService.getAll(PolicyType.PasswordGenerator);
+        : (await firstValueFrom(this.policyService.policies$)).filter(
+            (policy) => policy.enabled && policy.type === PolicyType.PasswordGenerator
+          );
```
This helper should probably be rewritten to expect an observable rather than grabbing the value as-needed

r-tome replied:
I've replaced this with an Observable function

MGibson1 commented on libs/common/src/services/policy/policy-api.service.ts:
```diff
@@ -79,7 +79,11 @@ export class PolicyApiService implements PolicyApiServiceAbstraction {
   }
 
-  async getPolicyForOrganization(policyType: PolicyType, organizationId: string): Promise<Policy> {
+  async getPolicyForOrganization(
+    policies: Policy[],
+    policyType: PolicyType,
+    organizationId: string
+  ): Promise<Policy> {
```
I don't see why this method belongs in this class atm, am I missing something? Feels like it's exclusively get logic from synced sources

MGibson1 commented on libs/common/src/services/policy/policy.service.ts:
```diff
@@ -13,43 +17,34 @@ import { ListResponse } from "../../models/response/listResponse";
   }
 
-  async getMasterPasswordPolicyOptions(policies?: Policy[]): Promise<MasterPasswordPolicyOptions> {
+  async getMasterPasswordPolicyOptions(policies: Policy[]): Promise<MasterPasswordPolicyOptions> {
```
seems weird to keep this particular method around when all the others have to go through the applies to user method.

r-tome replied:
This method is different to `policyAppliesToUser` because that one checks if the user has a specific policy on his organizations whereas this method returns the actual MasterPasswordPolicy info

MGibson1 reviewed (changes requested):
A few notes, the general feedback is to try using `pipe` and transformations to create new observables to be consumed in services and components rather than the service helper methods like `appliesToUser`.

Happy to discuss further if these suggestions aren't clear or if you've already explored this direction and found some issues with it.

r-tome commented on apps/web/src/app/organizations/manage/people.component.ts:
```diff
@@ -152,8 +165,14 @@ export class PeopleComponent
+  }
+
   async load() {
     const resetPasswordPolicy = await this.policyApiService.getPolicyForOrganization(
```
I removed the method `getPolicyForOrganization` from `policyApiService` and moved the code to this component as it was the only place using it.
I also updated the code to get `resetPasswordPolicy` through a subscription

r-tome commented on apps/web/src/app/settings/organization-plans.component.ts:
```diff
@@ -108,9 +121,23 @@ export class OrganizationPlansComponent implements OnInit {
+        takeUntil(this.destroy$),
+        concatMap(async (policies) => {
+          this.policies = policies;
+        })
```
I changed this to check `policyAppliesToUser` through a subscription.

r-tome commented on apps/web/src/app/tools/import-export/import.component.ts:
```diff
@@ -38,10 +41,22 @@ export class ImportComponent implements OnInit {
   async ngOnInit() {
     this.setImportOptions();
+    this.policyService.policies$
+      .pipe(
```
I've done a few local tests (debugging) and it all worked right. How would I go about testing this properly?

r-tome commented:
I've changed the Policy Service to return the Policies as an Observable collection and I've updated a few methods to also return Observable data. 
I wasn't able to refactor _everything_ to return an Observable because on some cases (classes that aren’t angular decorated) I can’t get around to getting the latest value from an Observable.

Hinton commented on apps/web/src/app/settings/two-factor-setup.component.ts:
```diff
@@ -26,7 +27,7 @@ import { TwoFactorYubiKeyComponent } from "./two-factor-yubikey.component";
   templateUrl: "two-factor-setup.component.html",
 })
 // eslint-disable-next-line rxjs-angular/prefer-takeuntil
```
Remove please.

Hinton commented on apps/web/src/app/settings/two-factor-setup.component.ts:
```diff
@@ -94,6 +98,18 @@ export class TwoFactorSetupComponent implements OnInit {
     this.providers.sort((a: any, b: any) => a.sort - b.sort);
     await this.load();
+
+    this.policyService
+      .policyAppliesToUser$(PolicyType.TwoFactorAuthentication)
+      .pipe(takeUntil(this.destroy$))
+      .subscribe((policyAppliesToUser) => {
+        this.twoFactorAuthPolicyAppliesToUser = policyAppliesToUser;
+      });
```
We should refactor this component to be responsive to the policy service. The logic in `evaluatePolicies` can probably be moved into this block.

r-tome replied:
`evaluatePolicies` is also called on `manage` that is triggered by a button click.

Hinton commented on apps/web/src/app/tools/import-export/import.component.ts:
```diff
@@ -38,10 +41,17 @@ export class ImportComponent implements OnInit {
 
   async ngOnInit() {
     this.setImportOptions();
+    this.policyService
+      .policyAppliesToUser$(PolicyType.PersonalOwnership)
+      .pipe(takeUntil(this.destroy$))
+      .subscribe((policyAppliesToUser) => {
+        this.importBlockedByPolicy = policyAppliesToUser;
+      });
```
I would argue that this code belongs in the template. However since we're also moving towards Reactive Forms we would have to change how that works almost immediately.

My suggestion is therefore to migrate the form, and remove the `this.importBlockedByPolicy`. The template can use the forms disabled state to show the warning.

Hinton commented on apps/web/src/app/vault/add-edit.component.ts:
```diff
@@ -24,7 +25,8 @@ import { LoginUriView } from "@bitwarden/common/models/view/loginUriView";
   templateUrl: "add-edit.component.html",
 })
-export class AddEditComponent extends BaseAddEditComponent {
+// eslint-disable-next-line rxjs-angular/prefer-takeuntil
```
Remove.

Hinton commented on apps/web/src/app/vault/add-edit.component.ts:
```diff
@@ -93,6 +96,16 @@ export class AddEditComponent extends BaseAddEditComponent {
+      .subscribe((options) => (this.options = options[0]));
+  }
+
+  // eslint-disable-next-line rxjs-angular/prefer-takeuntil
```
Remove.

Hinton commented on libs/common/src/services/state.service.ts:
```diff
@@ -736,24 +735,6 @@ export class StateService<
-  async getDecryptedPolicies(options?: StorageOptions): Promise<Policy[]> {
-    return (
-      await this.getAccount(this.reconcileOptions(options, await this.defaultInMemoryOptions()))
-    )?.data?.policies?.decrypted;
```
You should also update the `policies` in `AccountData` to `TemporaryDataEncryption`.

Hinton commented on libs/common/src/abstractions/policy/policy.service.abstraction.ts:
```diff
@@ -7,7 +9,7 @@ import { ListResponse } from "../../models/response/listResponse";
-  getAll: (type?: PolicyType, userId?: string) => Promise<Policy[]>;
+  policies$: Observable<Policy[]>;
 
   getMasterPasswordPolicyOptions: (policies?: Policy[]) => Promise<MasterPasswordPolicyOptions>;
```
Should be an observable

Hinton commented on libs/common/src/services/policy/policy.service.ts:
```diff
@@ -168,30 +171,28 @@ export class PolicyService implements InternalPolicyServiceAbstraction {
-        o.usePolicies &&
-        !this.isExcemptFromPolicies(o, policyType) &&
-        policySet.has(o.id)
+    return this.policies$.pipe(
```
Hmm, this won't work. In VaultTimeoutSettings service we rely on the userId. Since we're listening to a subscriber here the userId won't apply.

djsmith85 replied:
Would it be possible to turn the dependency of `VaultTimeoutSettingsService` on `PolicyService` around? 

Currently every time we call `VaultTimeoutService.shouldLock()`, we retrieve the vaultTimeout from `VaultTimeoutSettingsService.getVaultTimeout()` which then checks if the MaximumVaultTimeout-Policy applies to a given user. If it does, it will change the vaultTimeout and save it via the `stateService`. 

I don't see the need to check this that often, and a better way to check and enforce this policy, is to apply it once we have synced the policies.

- An active user will get the policy-update and then check if it needs to change the vaultTimeout based on the newly synced policy. 
- An inactive user (think account switching) will receive the policy update once they become the active user, executing a sync and a policy update. This will then update the vaultTimeout if necessary.

Rough idea
- `PolicyService` gets `VaultTimeoutSettingsService` injected into its ctor.
- `_policies` gets set 
- We check if the MaximumVaultTimeout policy applies to the current user (`activeAccountUnlocked`)
- If it does, then get the current vaultTimeout via `vaultTimeoutSettingsService.getVaultTimeout(userId)`
- Compare if currently set vaultTimeout needs to be updated based on the policy
- If it needs to be updated, call `vaultTimeoutSettingsService.setVaultTimeout(userId)`

This would be a performance improvement as we no longer unnecessarily check for policy changes every time we want to retrieve the vaultTimeout and it also breaks up the concerns.

Food for thought:
Ideally we'd break up `PolicyService` into individual policy-sub-services for each policy and they would listen to `PolicyService.polices$`. Each service would then check if their policy applies to the current user and takes the actions needed based on that policy. Baby steps though, we can tackle this in a future TDL PR.

Hinton replied:
Since this PR is getting kind of long drawn, perhaps we could cheat and have VaultTimeoutSettings get it's data some other way that isn't dependent on the observables, and add a new tech debt to clean this up. @r-tome

djsmith85 replied:
It's not cheating if we have good reason to take on tech-debt and we document it!

But...
I have had another look at this and I think we introduced an unintended change in behavior as we implemented account-switching.

- `policyAppliesToUser` was always used to check against the current user prior to account-switching
- With the intro of account-switching, we check all logged-in users if their vault should lock. 
- Which included passing in the userId, checking if the `MaximumVaultTimeout` policy applies to a given user.
- If it applies, then change the vaultTimeout and pass that value back to the `VaultTimeoutService` to check.

If we are going back to the old behavior -> Policy checks only apply to the currently active user. We can then remove the `userId` from `policyAppliesToUser()` and internally use the active userId instead. I'd also suggest renaming it to `policyAppliesToActiveUser()`.

This should be discussed with product though.

I still think the `PolicyService` or a policy-sub-service should take care of setting the `vaultTimeout` if the `MaximumVaultTimoutPolicy` gets synced/applied to a user. This can still be done in a future TDL PR.

@r-tome / @Hinton Feel free to ping me on other channels, if you'd like to discuss this and/or pair.

r-tome replied:
I've pushed changes aligned with @djsmith85's suggestions. Now `PolicyService` sets the `vaultTimeout` value on `VaultTimeoutSettings`.
This solution might have the caveat of only updating the policies for the current user but that is already an issue on using the `policies$` observable.
@Hinton let me know what you think.

r-tome replied:
I've reverted the changes back to the `VaultTimeoutSettings` as it was introducing an unintended change and @Hinton is right, this PR is drawn enough as is. Now `VaultTimeoutSettings` gets the policies directly as it was before.

Hinton reviewed (commented):
Please update the subscribe pattern.

r-tome commented on libs/angular/src/components/update-temp-password.component.ts:
```diff
@@ -54,13 +54,16 @@ export class UpdateTempPasswordComponent extends BaseChangePasswordComponent {
+  ngOnDestroy() {
+    super.ngOnDestroy();
+  }
+
```
Right, I've removed it.

Hinton commented on libs/common/src/abstractions/state.service.ts:
```diff
@@ -104,6 +105,8 @@ export abstract class StateService<T extends Account = Account> {
   getDecryptedPinProtected: (options?: StorageOptions) => Promise<EncString>;
   setDecryptedPinProtected: (value: EncString, options?: StorageOptions) => Promise<void>;
+  getDecryptedPolicies: (options?: StorageOptions) => Promise<Policy[]>;
+  setDecryptedPolicies: (value: Policy[], options?: StorageOptions) => Promise<void>;
```
Mark these as deprecated please.

Hinton commented on libs/common/src/services/policy/policy.service.ts:
```diff
@@ -45,6 +45,28 @@ export class PolicyService implements InternalPolicyServiceAbstraction {
+    } else {
+      const diskPolicies = await this.stateService.getEncryptedPolicies({ userId: userId });
+      for (const id in diskPolicies) {
+        // eslint-disable-next-line
```
Please use the prototype which eslint warns about.

Hinton commented on libs/common/src/services/policy/policy.service.ts:
```diff
@@ -233,4 +252,26 @@ export class PolicyService implements InternalPolicyServiceAbstraction {
+    policies: Policy[],
+    policyType: PolicyType,
+    policyFilter: (policy: Policy) => boolean = (p) => true,
+    userId?: string
```
Remove UserId

Hinton reviewed (dismissed):
Please ensure QA regression test everything related to policies, account switching, vault locking.

djsmith85 commented:
@r-tome Awesome, work on this 👍

r-tome commented:
@MGibson1 are you ok with this being merged? It's gone through QA and passed all tests. I can't merge it because you still have a change request.
